keps/sig-multicluster/2149-clusterid/README.md

```diff
@@ -0,0 +1,705 @@
+of this KEP is to provide a flexible but useful baseline for clusterID that can
+work with the known use cases.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+* Propose a standard for how cluster identification metadata should be stored and 
+managed as Kubernetes resources
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+* Define any characteristics of the system that tracks cluster IDs within a cluster (i.e. a cluster registry)
+* Solve any problems without specific, tangible use cases (though we will leave room for extension).
+
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### Overview
+Each cluster in a ClusterSet will be assigned a unique identifier that lives at least as long as its membership in the set, and is immutable for its lifetime. This identifier will be stored in a new cluster-scoped `ClusterClaim` CR with the well known name `id.k8s.io` that may be referenced by workloads within the cluster. The identifier must be a valid [RFC-1123](https://tools.ietf.org/html/rfc1123) DNS label, and may be created by an implementation dependent mechanism.
+
+While a member of a ClusterSet, a cluster will also have an additional `clusterset.k8s.io ClusterClaim` which describes its current membership. This claim must be present exactly as long as the set membership lasts, and removed when the cluster is no longer a member.
+
+### User Stories
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### ClusterSet membership
+
+I have some set of clusters working together and need a way to uniquely identify them within the system that I use to track membership.
+
+#### Multi-Cluster Services
+
+I have a headless multi-cluster service deployed across clusters in my ClusterSet with similarly named pods in each cluster. I need a way to disambiguate each backend pod via DNS.
+
+#### Diagnostics
+
+Clusters within my ClusterSet send logs/metrics to a common monitoring solution and I need to be able to identify the cluster from which a given set of events originated.
+
+
+#### Joining or moving between ClusterSets
+
+I want the ability to add a previously-isolated cluster to a ClusterSet, or to move a cluster from one ClusterSet to another and be aware of this change.
+
```

Comment:
I do not think we should allow that.  The complexity of it spirals very quickly.  All use-cases I have seen so far can be solved better in other ways.  Curious to hear yours.

Reply:
It would be good to discuss this. A few of the assumptions we've been stating around clustersets (e.g. namespace sameness, MCS import behavior) have some sppoky action at a distance if a cluster can belong to multiple clustersets. I think from cases we've discussed it would be safer to have a single clusterset with some way to create internal separation where needed

Reply:
Put this as a nongoal for now explicitly in 3fa5623c2886ebd4bc829d3a04aac189b09baeae, but at the last SIG-Multicluster meeting that was some chat about this (also by the phrase "clusterset collaboration") and I am definitely interested in more viewpoints/use cases on this!